v2/controller.go

```diff
@@ -0,0 +1,212 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package v2
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	statsv2 "github.com/containerd/cgroups/v2/stats"
+	"github.com/pkg/errors"
+)
+
+type cgValuer interface {
+	Values() []CGStr
+}
+
+type Resources struct {
+	CPU    *CPU
+	Memory *Memory
+	Pids   *Pids
+}
+
+func (r *Resources) Values() (o []CGStr) {
+	values := []cgValuer{
+		r.CPU,
+		r.Memory,
+		r.Pids,
+	}
+	for _, v := range values {
+		if v == nil {
+			continue
+		}
+		o = append(o, v.Values()...)
+	}
+	return o
+}
+
+type CGStr struct {
+	key   string
+	value interface{}
+}
+
+func (c *CGStr) write(path string, perm os.FileMode) error {
+	var data []byte
+	switch t := c.value.(type) {
+	case uint64:
+		data = []byte(strconv.FormatUint(t, 10))
+	case int64:
+		data = []byte(strconv.FormatInt(t, 10))
+	case []byte:
+		data = t
+	case string:
+		data = []byte(t)
+	default:
+		return ErrInvalidFormat
+	}
+	return ioutil.WriteFile(
+		filepath.Join(path, c.key),
+		data,
+		perm,
+	)
+}
+
+func writeCGStr(path string, strs []CGStr) error {
+	for _, o := range strs {
+		if err := o.write(path, defaultFilePerm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func NewController(mountpoint string, g string, resources *Resources) (*Controller, error) {
```

Comment:
Same, since `g` not any more of the GroupPath type, it's not so obvious what it stands for.